Add contribution guide (#629)
neondatabase/website: content/docs/community/contribution-guide.md

---

This page provides guidelines for contributing to the Neon documentation. Our goal is to create an environment where our community has the information and knowledge required to confidently participate in improving the Neon documentation.

## Why should you contribute?

Open-source projects are always evolving. Contributing to documentation is a great way for beginners to get started in open source and for experienced developers to explain complex topics while sharing their knowledge with the community.

By contributing to the Neon docs, you're helping us create a stronger learning resource for all developers. Whether you've found a typo, a section that's hard to understand, or you've noticed that a certain topic is missing, your contribution is always welcome and appreciated.

## How to contribute

The content for the documentation is located in the [neondatabase/website](https://github.com/neondatabase/website) repository, in the `/content/docs` directory. To contribute, you have two options:

1. Edit files directly on GitHub.
2. Fork the [neondatabase/website](https://github.com/neondatabase/website) repository, create a branch for your changes, and submit a pull request.

> If you prefer the first option, which is great for edits and small updates, there is an **Edit this page** link at the bottom of each Neon documentation page.

Reviewer: FYI, I noted in the "Doc portal" [Notion page ](https://www.notion.so/neondatabase/Docs-Portal-Design-Phase-3-13ccf42119bb4a7c8697ecce2b44cfb6) that we might want to move this Edit link (or a simpler icon) to the top of the page at some point, if we want to encourage more contributors.

![GitHub edit this page link](/docs/community/edit_this_page.png)

Clicking the link takes you to the Markdown file in GitHub, where you can click the **Edit this page** icon to make a change. When you finish editing, commit your changes to create a pull request.

If you would rather fork the [neondatabase/website](https://github.com/neondatabase/website) repository and submit pull requests, but you're not familiar with the process, we suggest going through the [GitHub Open Source Guide](https://opensource.guide/how-to-contribute/#opening-a-pull-request). This guide describes how to fork a repository, create a branch, and submit a pull request.

## Markdown

> Neon uses Markdown as the documentation source format. Markdown is a lightweight markup language that lets you add formatting elements to plaintext text documents. It's designed to be easy to read and write.

Reviewer: ```suggestion
Neon uses Markdown as the documentation source format. Markdown is a lightweight markup language that lets you add formatting elements to plaintext text documents. It's designed to be easy to read and easy to write.
```

Reviewer: A small rhetorical flourish: you can take or leave it.

## Documentation table of contents

This section describes how to modify the documentation table of contents, also referred to as the "sidebar". Adding, removing, or moving a page in the documentation requires updating the sidebar. The sidebar is defined in a `yaml` file, conveniently named `sidebar.yaml`, which you can find at the root of the `/docs` directory.

### Add a new category

To add a new category to the sidebar, add a new item to the top-level array with `title` and `items` key values, as shown below:

```diff
 - title: Category 1
   items:
     - title: Page 1
       slug: page-1
+- title: Category 2
+  items:
+    - title: Page 2
+      slug: page-2
```

### Add a new page

To add new page, add a new item to the `items` array with the `title` and `slug` keys under the category or subcategory.

```diff yaml
 - title: Category 1
   items:
     - title: Page 1
       slug: page-1
 - title: Category 2
   items:
     - title: Page 2
       slug: page-2
    - title: Subcategory 1
      items:
        - title: Page 3
          slug: page-3
+       - title: Page 4
+         slug: page-4
    - title: Page 5
      slug: page-5
```

> - The `title` in the sidebar may differ from `title` in the Markdown file. For example, your sidebar title might be a shorter version of the title in your Markdown file.

Reviewer: ```suggestion
- The `title` in the sidebar may differ from `title` in the Markdown file. For example, your sidebar title might be a shorter version of the title in your Markdown file. This lets you write longer, more informative page titles while keeping the sidebar titles short, readable, and easy to scan. These titles should remain logically related. For example, in our docs we reduce the page title "Use Grafbase Edge Resolvers with Neon" to just "Grafbase" in the sidebar.
```

> - `slug` should always match page's slug.

Reviewer: ```suggestion
- `slug` should always exactly match the page's slug (the last part of the URL after the final backslash "/", in our case the name of the Markdown file).
```

## Markdown frontmatter

> Each Neon documentation Markdown file includes frontmatter section at the beginning of the file containing file metadata. The frontmatter section is distinguished by three dashes, as shown here:

Reviewer: ```suggestion
Each Neon documentation Markdown file includes a frontmatter section at the beginning of the file containing file metadata. The frontmatter section is distinguished by three dashes, as shown here:
```

Neon documentation same page link: [Code blocks](#code-blocks)
Italics markup: _italic_
Bold markup: **strong**
monospace: `backtick`
```

## Code blocks

To insert a code block into your Markdown file, specify three backticks (```) on the lines before and after the code. Specify the language identifier to enable code highlighting, as in this example:

````md
```sql
SELECT * FROM posts ORDER BY id;
```
````

You can add language-specific highlighting to code blocks, as in the example above. See [Supported language highlighting for code blocks](https://prismjs.com/index.html#supported-languages).

You can also wrap code blocks in `<CodeBlock></CodeBlock>` tags to show line numbers or wrap long strings.

`<CodeBlock>` accepts the following options:

- `showLineNumbers`: Shows the line numbers in the code block
- `shouldWrap`: Enables code wrapping in the code block. This is useful for long commands or connection strings that would otherwise require the reader to scroll.

Example:

````md
<CodeBlock shouldWrap>

```text
DATABASE_URL=postgres://alex:AbC123dEf@ep-cool-darkness-123456.us-east-2.aws.neon.tech/dbname?connect_timeout=15&pool_timeout=15`
```

</CodeBlock>
````

> 

Reviewer: Does shouldWrap not work in preview? It's not wrapping for me on my browser (mac chrome):
![image](https://github.com/neondatabase/website/assets/5885546/d6558541-0325-4b47-9ea3-c92c63486c27)

Author: This example shows what the markdown looks like with the tags, so the line would not be wrapped here, but I've replaced the line to avoid any confusion.

Diagrams and images are stored in the `/public/docs` directory in the Neon website repository. The directory location of the diagram or image under `public` mirrors the location of the file that includes the diagram or image, as shown below:

Example file structure:

```md
├── content
│ ├── docs
│   ├── introduction
│     ├── architecture-overview.md

├── public
│ ├── docs
│   ├── introduction
│     ├── neon_architecture.png // put images in the public directory with the same name
```

To add an image to your Markdown file, add an entry that looks like this:

```md
![Neon architecture diagram](/docs/introduction/neon_architecture.png)
```

## Style Guide

This section outlines the stylistic elements that we do our best to follow in the Neon documentation.

### Voice and language

The voice in the documentation should sound like one human being explaining something to another, while striking the right balance between being approachable and professional.

#### Guidelines

1. **Use contractions**:
    - **Do**: Use contractions like "it's", "don't", "you're" to make the tone more conversational.
      - *Example*: "It's essential to save your progress."

>     - **Don't**: Overuse contractions, which can compromise clarifyS.

Reviewer: ```suggestion
    - **Don't**: Overuse contractions, which can compromise clarity.
```

2. **Simplicity over jargon**:
    - **Do**: Choose simpler words when possible.
      - *Example*: "Use the tool," not "Utilize the instrument."

>     - **Don't**: Oversimplify to the point of being inaccurate.

Reviewer: ```suggestion
    - **Don't**: Oversimplify to the point of being inaccurate or leaving out useful context.
```

3. **Active voice**:
    - **Do**: Prefer active voice.
      - *Example*: "The software converts the file."
    - **Don't**: Over-rely on passive voice.
      - *Example*: "The file is converted by the software."

4. **Brief sentences**:
    - **Do**: Keep sentences concise.
      - *Example*: "Check the settings."

5. **Personalize when relevant**:
    - **Do**: Use "you" to address the reader.
      - *Example*: "You can adjust the setting."
    - **Don't**: Overdo direct addresses. Not every sentence should start with "You".

6. **Consistent terminology**:
    - **Do**: Stick to one term for one concept.

>       - *Example*: Always use "dashboard", not sometimes "control panel".

Reviewer: ```suggestion
      - *Example*: Always use "dashboard". Don't mix that term with "control panel".
```

Follow the capitalization style used by the authoritative source, which may use non-standard case styles. For example: PostgreSQL, GitHub, npm.

### Fake user information

When including user information in connection details, API calls, or UI instructions, don't use real user information or email addresses.

- Use an email address ending in `@example.com` or `domain.com`.
- Use strings like `example_username` or one of the following diverse and non-gendered names: Zhang Kai, Alex Lopez, or Dana Smith.

### Connection strings

Connection strings should be defined as follows:

```text
postgres://[user]:[password]@[neon_hostname]/[dbname]
```

If you need to provide a connection string with realistic values, use one of the user names mentioned above, `AbC123dEf` for the password, and `dbname` for the database name:

```text
postgres://alex:AbC123dEf@ep-cool-darkness-123456.us-east-2.aws.neon.tech/dbname
```

### Commands, parameters, values, filenames

Commands, parameters, values, filenames, error messages, connection strings, and other similar items should be enclosed in backticks. For example:

- "Run the `neonctl projects list` command."

> - "Execute 'git clone' to clone a Git repository..."

Reviewer: ```suggestion
- "Execute `git clone` to clone a Git repository..."
```

> - 'git clone' is a command that should be in lowercase, whereas Git is the product and should have a capital G.

Reviewer: ```suggestion
- `git clone` is a command that should be in lowercase, whereas Git is the product and should have a capital G.
```

- "A connection string has this format: `postgres://[user]:[password]@[neon_hostname]/[dbname]`"

## Questions?

> If you have questions or run into any issues, please reach out to us on the [Neon Community forum](https://community.neon.tech/).

Reviewer: If they are docs process questions, will putting those to the community get to the docs team?

Author: I think this is OK. If someone asks a question about doc processes and we don't see it right away, I think someone on the team will let us know.
